### DISCUSSION TITLE
Moving to Mason

### DISCUSSION THREAD

**Discussion by mehalter** [AstroNvim/AstroNvim#820] [Announcements]
With PR #793 merged into the `main` and available on AstroNvim `nightly` update channel we have officially started the process of  moving to Mason. Due to the urgency of this update in order to keep the ability to update language servers, we have formed it in a way to be non-breaking with current user configurations. You will need to reinstall language servers after the update. If you have language servers defined in your config as `ensure_installed` then these will be automatically installed after the first time loading AstroNvim with the Mason plugin installed.

Hopefully this doesn't cause a lot of friction and I will wait a bit before we tag the next stable AstroNvim release to give time for people to submit any bugs or concerns.

*Note*: We have left the compatibility with the `plugins.nvim-lsp-installer` table for configuring the language servers that will be installed alongside the new `plugins.mason-lspconfig` table. This backwards compatibility will be getting deprecated in `v2.0.0` so we would recommend that you rename your `nvim-lsp-installer` configuration to `mason-lspconfig` whenever you get the chance in preparation.

New Plugins:

- [mason](https://github.com/williamboman/mason.nvim) - The new package manager to replace `nvim-lsp-installer`
- [mason-lspconfig](https://github.com/williamboman/mason-lspconfig.nvim) - This will be used to automatically install language servers for `lspconfig` and automatically set them up appropriately
- [mason-tool-installer](https://github.com/WhoIsSethDaniel/mason-tool-installer.nvim) - This will be used to automatically install formatters and linters that will be used with `null-ls`

**Comment by mehalter** [comment]
If anyone does not want to use Mason. Feel free to use the following `updater` settings in your `user/init.lua` file which will pin your AstroNvim to the commit directly before the Mason adoption:

```lua
updater = {
  channel = "nightly",
  commit = "5136e0f94d344aa76a826c77c81056082e5b34dd",
},
```

**Comment by sbeyer** [comment]
It actually broke my user configuration for rust-tools. Haven't used any Rust recently so I just deactivated my user configuration for that and haven't looked for a solution so far.

EDIT: FYI the deactivated code is this inside my `plugins/init.lua`:
```lua
  -- Rust goodies
  {
    'simrat39/rust-tools.nvim',

    requires = {
      "nvim-lspconfig",
      "nvim-lsp-installer",
      -- "nvim-dap",
    }
  },
```
and the file `lsp/server_registration.lua` with this code:
```lua
return function(server, opts)
  -- Integration of rust-tools
  -- from https://github.com/williamboman/nvim-lsp-installer/wiki/Rust
  if server.name == "rust_analyzer" then
    -- Initialize the LSP via rust-tools instead
    require("rust-tools").setup {
        -- The "server" property provided in rust-tools setup function are the
        -- settings rust-tools will provide to lspconfig during init.            -- 
        -- We merge the necessary settings from nvim-lsp-installer (server:get_default_options())
        -- with the user's own settings (opts).
        server = vim.tbl_deep_extend("force", server:get_default_options(), opts),
    }
    server:attach_buffers()
    -- Only if standalone support is needed
    require("rust-tools").start_standalone_if_required()
  else
    server:setup(opts)
  end
end
```
(I haven't even checked if it is sufficient to fix the `requires =` in `plugins/init.lua`.)

**Reply by mehalter** [reply]
@sbeyer check out the documentation for a cleaner way to set up `rust-tools`: https://astronvim.github.io/Recipes/advanced_lsp#lsp-specific-plugins

**Reply by mehalter** [reply]
Ah I just realized I haven't added `rust-tools` example to that page specifically. It should.look something like this:

```lua
return {
  lsp = {
    skip_setup = { "rust_analyzer" }, -- skip lsp setup because rust-tools will do it itself

  },
  plugins = {
    init = {
      {
        "simrat39/rust-tools.nvim",
        after = "mason-lspconfig.nvim", -- make sure to load after mason-lspconfig
        config = function()
          require("rust-tools").setup {
            server = astronvim.lsp.server_settings "rust_analyzer", -- get the server settings and built in capabilities/on_attach
          }
        end,
      },
    },
    ["mason-lspconfig"] = {
      ensure_installed = { "rust_analyzer" }, -- install rust_analyzer
    },
  },
}
```

**Reply by sbeyer** [reply]
Works like a charm! :D Also it's good that we talked about it, because I'm also using the clangd-extensions config now.
